mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScanner.java

```diff
@@ -0,0 +1,191 @@
+import com.datastax.oss.driver.shaded.guava.common.collect.Sets;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Types;
+
+/** Provides mechanisms for building and traversing a class/interfaces hierarchy. */
+public class HierarchyScanner {
+
+  // placeholder class for resolving the default HierarchyScanStrategy
+  @HierarchyScanStrategy
+  private static final class ClassForDefaultScanStrategy {}
+
+  /**
+   * Resolves the type hierarchy for the given type by first looking for a {@link
+   * HierarchyScanStrategy}-annotated class or interface in the given types hierarchy. The hierarchy
+   * is traversed until a {@link HierarchyScanStrategy} is encountered in a depth-first manner as
+   * follows:
+   *
+   * <ol>
+   *   <li>Initialize <code>i</code> as an empty set
+   *   <li>Visit typeElement (<code>e</code>)
+   *   <li>Visit <code>e</code>'s parent class (<code>p</code>), if <code>p</code> is null, stop
+   *       traversing after evaluating remaining interfaces
+   *   <li>Visit <code>e</code>'s interfaces, and record those interfaces' parents for later use (
+   *       <code>ni</code>)
+   *   <li>Visit <code>i</code>, and append those interface's parents to <code>ni</code> for later
+   *       use)
+   *   <li>If <code>p != null</code> </code>Set <code>e := p, i := ni</code> and repeat starting at
+   *       step 3
+   *   <li>Visit <code>ni</code> interfaces and their parents until we've reached root
+   * </ol>
+   *
+   * Once a {@link HierarchyScanStrategy} is identified, the returning hierarchy is built by
+   * traversing the hierarchy again using the chosen strategy.
+   *
+   * @param typeElement The type whose hierarchy will be traversed.
+   * @param types Used to resolve parent type elements
+   * @return The type hierarchy, ordered from typeElement to the highestAncestor, as dictated by the
+   *     resolved {@link HierarchyScanStrategy}
+   */
+  public static Set<TypeElement> resolveTypeHierarchy(TypeElement typeElement, Types types) {
+    HierarchyScanStrategy hierarchyScanStrategy =
+        HierarchyScanner.resolveHierarchyScanStrategy(typeElement, types);
+
+    ImmutableSet.Builder<TypeElement> hierarchy = ImmutableSet.builder();
+    traverseFullHierarchy(hierarchyScanStrategy, typeElement, types, hierarchy::add);
+    return hierarchy.build();
+  }
+
+  private static HierarchyScanStrategy resolveHierarchyScanStrategy(
+      TypeElement classElement, Types types) {
+    // Use the default HierarchyScanStrategy to find the configured HierarchyScanStrategy.
+    // This is done because the default strategy is the most permissive.
+    HierarchyScanStrategy strategy =
```

Comment:
possible optimization could be done here such that if the strategy is not defined, return the fully traversed hierarchy so we don't have to traverse it again...but it's probably such a cheap operation that is only done for each dao/entity that it might not be worth it.

Reply:
Decided it's not really worth it, would require changing the API a bit and at this point things work great...